sharedbuffers/mapped_struct.py

```diff
@@ -317,113 +323,331 @@ def unpack_from(cls, buf, offs, idmap = None):
+            dataoffs += 8
+            return dcode, objlen, 8, dataoffs
 
-    @classmethod
-    def unpack_from(cls, buf, offs, idmap = None, array = array.array, itemsizes = {
-                dtype : array.array(dtype, []).itemsize
-                for dtype in ('B','b','H','h','I','i','l','d')
-            } ):
-        if idmap is None:
-            idmap = {}
-        if offs in idmap:
-            return idmap[offs]
+        else:
+            raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
 
-        buf = _likerobuffer(buf)
-        dcode = buf[offs]
-        dataoffs = offs
+    def _metadata(self,
+        itemsizes =  {dtype : array.array(dtype, []).itemsize for dtype in ('B','b','H','h','I','i','l','d')},
+        ):
+
+        dataoffs = self.offs
+        buf = self.buf
+        dcode = buf[dataoffs]
 
         if dcode in ('B','b','H','h','I','i'):
             objlen, = struct.unpack('<I', buf[dataoffs+1:dataoffs+4] + '\x00')
             dataoffs += 4
             if objlen == 0xFFFFFF:
                 objlen = struct.unpack_from('<Q', buf, dataoffs)
                 dataoffs += 8
-            rv = cls(buf, offs, dataoffs, dcode, itemsizes[dcode], objlen, idmap)
+            return dcode, objlen, itemsizes[dcode], dataoffs, struct.Struct(dcode)
+
         elif dcode == 'q':
             objlen, = struct.unpack('<Q', buf[dataoffs+1:dataoffs+8] + '\x00')
             dataoffs += 8
-            rv = cls(buf, offs, dataoffs, dcode, itemsizes['l'], objlen, idmap)
+            return dcode, objlen, itemsizes['l'], dataoffs, struct.Struct('l')
+
         elif dcode == 'd':
             objlen, = struct.unpack('<Q', buf[dataoffs+1:dataoffs+8] + '\x00')
             dataoffs += 8
-            rv = cls(buf, offs, dataoffs, dcode, itemsizes[dcode], objlen, idmap)
+            return dcode, objlen, itemsizes['d'], dataoffs, struct.Struct('d')
+
         elif dcode == 't':
             objlen, = struct.unpack('<Q', buf[dataoffs+1:dataoffs+8] + '\x00')
             dataoffs += 8
-            rv = cls(buf, offs, dataoffs, dcode, itemsizes['l'], objlen, idmap)
+            return dcode, objlen, itemsizes['l'], dataoffs, struct.Struct('l')
+
         else:
             raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
-        return rv
 
-    @cython.locals(obj_offs = cython.ulonglong, dcode = object, index=cython.ulonglong)
+    def __init__(self, buf, offs, idmap = None):
+        self.offs = offs
+        self.buf = buf
+        self.idmap = {} if idmap is None else idmap
+
+        if cython.compiled:
+            self._cinit()
+            self._cmetadata()
+        else:
+            # Call metadata to check the object
+            self._metadata()
+
+    @cython.ccall
+    def _cinit(self):
+        self.pybuf.buf = cython.NULL
+        PyObject_GetBuffer(self.buf, cython.address(self.pybuf), PyBUF_SIMPLE)  # lint:ok
```

Comment:
Beware, if you do this, you need a `__dealloc__`

Reply:
I am using __del__ for releasing the buffer. Is it no ok?

Reply:
It's not the same really. Check out cython docs about it.